Histogram APIs for device policies

danhoeflinger opened the pull request:
Creates two new APIs for histogram for use with device policies: 
---

Evenly divided bins:
```
template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _T, typename _RandomAccessIterator2>
oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _RandomAccessIterator2>
histogram(_ExecutionPolicy&& policy, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last, const _Size& __num_bins,
          const _T& __first_bin_min_val, const _T& __last_bin_max_val, _RandomAccessIterator2 __histogram_first);
```
Creates `__num_bins` evenly divided bins between `__first_bin_min_val` and `__last_bin_max_val`. 

Custom bin boundary ranges:
```
template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _RandomAccessIterator2, typename _RandomAccessIterator3>
oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _RandomAccessIterator3>
histogram(_ExecutionPolicy&& policy, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last, _RandomAccessIterator2 __boundary_first,
          _RandomAccessIterator2 __boundary_last, _RandomAccessIterator3 __histogram_first);
```
Creates bins in between each boundary provided in the range `__boundary_first` to `__boundary_last`.   


Both APIs set all output bins to zero, then count the input elements which fall into each bin, safely ignoring input elements which do not fall within any bin.  Output bin data must be allocated and passed in by caller.  Returns iterator to the end of the output histogram bin range. 

---

This PR adds the corresponding tests for these two histogram APIs.  
Currently these APIs do not support host side policies for any backend, this will be added in a separate PR.

SergeyKopienko commented on test/parallel_api/numeric/numeric.ops/histogram.pass.cpp:
```diff
@@ -0,0 +1,131 @@
+        }
+
+        // testing of large number of items may take too much time in debug mode
+        Size n =
```
constexpr?

SergeyKopienko replied:
Still see `Size n =` in diffs...

danhoeflinger replied:
hmm, make sure you are looking at all commits rather than just a subset... it is changed in this commit: https://github.com/oneapi-src/oneDPL/pull/1273/commits/5cc254c88860f476512e74877cb7d7b5cbb4605f

SergeyKopienko commented on test/parallel_api/numeric/numeric.ops/histogram.pass.cpp:
```diff
@@ -0,0 +1,131 @@
+    {
+        const Size bin_size = bin_last - bin_first;
+        histogram_sequential(in_first, in_last, bin_size, bin_min, bin_max, expected_bin_first);
+        auto orr = ::oneapi::dpl::histogram(exec, in_first, in_last, bin_size, bin_min, bin_max, bin_first);
```
Where is defined `::oneapi::dpl::histogram` ?
Looks like it's defined in the file `include/oneapi/dpl/internal/histogram_extension_defs.h'
It's incorrect, because as far as I understand we shouldn't declare public algorithms inside internal folder.

danhoeflinger replied:
This is a good point... I followed the lead of the other extension algorithms `exclusive_scan_by_segment`, async routines, etc.

Do we need to add a new top level header like `include/oneapi/dpl/algorithm` for these extensions and another for  experimental APIs?  Right now they are included when one includes `algorithm` but defined in `include/oneapi/dpl/internal/`.

SergeyKopienko replied:
I don't know exactly the right way, May be better to discuss with other colleagues.

danhoeflinger replied:
I think we need to consider it with the experimental async algorithms, and the other extensions:  

`include/oneapi/dpl/internal/binary_search_extension_defs.h`
`include/oneapi/dpl/internal/async_extension_defs.h` @reble 
`include/oneapi/dpl/internal/by_segment_extension_defs.h` @mmichel11 

@akukanov, @rarutyun: do you have an opinion here on where these extensions should be defined?

danhoeflinger replied:
After talking to @timmiesmith , he mentioned this was originally done this way to separate extensions what was being upstreamed to PSTL.  I think perhaps all of these extensions will end up in a compat or extensions folder eventually.

mmichel11 commented on include/oneapi/dpl/internal/histogram_extension_defs.h:
```diff
@@ -0,0 +1,48 @@
+{
+
+// Only supported for dpcpp backend
+#ifdef _ONEDPL_BACKEND_SYCL
```
Since at the moment we are only supporting the SYCL backend use case, would it be beneficial to have some sort of `static_assert` or graceful error if the user attempts to compile for a non-SYCL backend to prevent confusion?

danhoeflinger replied:
I looked at doing this (a few commits back in the history).  However, for some compilers it seems the function is instantiated and hits the `static_assert` even without an actual call of this function.  We could instead do it with a runtime assert, perhaps that is the way to go.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,550 @@
+        ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+        ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__bin_hash.__boundaries.size(), __gSize);
+        ::std::uint8_t __k;
+        __boundary_type* __d_boundaries = (__boundary_type*)(__boost_mem);
```
Yes, thanks.  Done.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,550 @@
+                __reduce_out_histograms<_bin_type>(__local_histogram, 0, __bins, __num_bins, __self_item);
+            });
+    });
+    __e.wait();
```
Sure, I've converted it to return a future up to the `__parallel_histogram` call and wait on it there.  It should be easy to add an async version.

MikeDvorskiy commented on include/oneapi/dpl/internal/histogram_extension_defs.h:
```diff
@@ -0,0 +1,48 @@
+// -*- C++ -*-
```
Why do we put public API to internal folder?
I would  except to see 
Public API for histogram in `.../dpl/pstl/glue_algorithm_def.h `
Host Pattern - in `.../dpl/pstl/algorithm_impl.h `
hetero Pattern  - in `.../dpl/pstl/hetero/algorithm_impl_hetero.h `

MikeDvorskiy commented on include/oneapi/dpl/internal/histogram_extension_defs.h:
```diff
@@ -0,0 +1,48 @@
+          typename _RandomAccessIterator2>
+oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _RandomAccessIterator2>
+histogram(_ExecutionPolicy&& policy, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+          const _Size& __num_bins, const _T& __first_bin_min_val, const _T& __last_bin_max_val,
```
Usually, there is practice (according to std signatures) to pass  Size/count by value.

danhoeflinger replied:
1.  Sure, I'll make that change.  
2. Sure, we can do that.  I suppose its maybe better to convert to the value type here than to wait until the comparisons to reconcile types.

I'll update when this is done.

danhoeflinger replied:
@MikeDvorskiy, do you think we should also pass `__first_bin_[min/max]_val` by value as well? In most cases its probably better.

danhoeflinger replied:
I've converted to have all the scalar* inputs to be passed by value and utilized the `value_type`.  Please take a look to see if this is what you had in mind.  
I've converted `_T` to be a defaulted template parameter (`_ValueType`) which is default equal to `iterator_traits<_RandomAccessIterator1>::value_type`.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+    }
+
+    inline void
+    init_SLM_memory(void* __boost_mem, const sycl::nd_item<1>& __self_item) const
```
`void*` and further `static_cast` technic doesn't look a C++ type safety approach... Probably, can we avoid that?

danhoeflinger replied:
My intention here is to allow `__binhash_SLM_wrapper<T>` to be called generically by the kenels without needing to know how the extra SLM memory is going to be used by the individual bin hash structure.  For evenly divided bins, it is unused (and size=0).  For custom ranges, it is used to cache the boundaries.  In theory, one could use the SLM memory for something else if they add another histogram API in the future.  

Is there a better way than void* to handle this memory but keep it a generic call?

danhoeflinger replied:
I've refactored this a bit, letting binhash dictate the type of the extra memory.  I've also split out the extra memory to a separate block from the local histogram bin storage.  This allows for better code and not having to cast pointers around.

There may be a minor performance hit by splitting the SLM allocations this but it is better code.  I'll investigate to make sure we don't lose too much.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+    ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+    ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+    ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__num_bins, __gSize);
+    ::std::uint8_t __k;
```
uninitialized variable

danhoeflinger replied:
It is initialized in the for loop directly below, but I can move the initialization up a line.  Its a safety vs clarity choice I think.

MikeDvorskiy replied:
Usually,  a code analyzing tool treats a variable definition without initialization as an issue....

danhoeflinger replied:
Thanks, I've edited this to initialize on definition.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+inline auto
+__parallel_histogram_sycl_impl(_ExecutionPolicy&& exec, _Iter1 __first, _Iter1 __last, _Iter2 __histogram_first,
+                               const _Size& __num_bins, _IdxHashFunc __func, _Range&&... __opt_range)
+{
```
I guess, it makes sense to process trivial cases when __num_bins = 0 or 1.

danhoeflinger replied:
I do the specialization for `__num_bins == 0` higher in the stack.  

With the semantics we are using, 1 is not actually a trivial case, because input elements which go outside of the specified bin range are ignored.  Therefore, you can't merely put all elements into the single bin because they all may not map there.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+    auto bins_buf = keep_bins(__histogram_first, __histogram_first + __num_bins);
+
+    auto __f = oneapi::dpl::__internal::fill_functor<__global_histogram_type>{__global_histogram_type{0}};
+    //fill histogram bins with zeros
```
we submit additional kernel here.. Probably, is it possible to fill histogram bins with zeros within "main" kernel, which computes histogram?

danhoeflinger replied:
Sadly, I don't think its possible because we cant have a cross-workgroup barrier to synchronize.  No input elements can be processed until all output histogram elements have been zeroed, otherwise there is a race, because we don't know where the input elements will map to.

If we want to optimize for a single workgroup version, we could combine these kernels, but otherwise I don't see a way.

MikeDvorskiy replied:
It is global barrier required.. I see.
BTW,  if we create an write accessor without no_init property... and we have non-implicit-lifetime type T, which T{} gives zero?

danhoeflinger replied:
After thinking about this further.  I don't think this benefits us.  
If we create a separate "known type" copy of the output histogram to take advantage of the T{} zero, then we still need a global barrier after all workgroups are done accumulating to this output histogram before we can copy to the final user provided output data.  This would merely trade a zeroing kernel before the main kernel for a copying kernel after, and add an extra copy of the histogram data to global memory.  

Also, I expect coordinated zeroing of local histograms in the kernel to be faster than a data transfer of default constructed data from the host.  (`read_write` + `no_init` +  explicit zeroing) vs (`read_write` + default constructed zero).  

@MikeDvorskiy Let me know your thoughts here, and tell me if I'm missing something.

MikeDvorskiy commented on test/parallel_api/numeric/numeric.ops/histogram.pass.cpp:
```diff
@@ -0,0 +1,131 @@
+#    if PSTL_USE_DEBUG
+            70000000;
+#    else
+            100000000;
```
`1e8` ? just for more readability)

SergeyKopienko replied:
or `100'000'000;`

danhoeflinger replied:
Switched to scientific notation, thanks.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+
+    template <typename _T2>
+    inline ::std::uint32_t
+    get_bin(_T2&& __value, void* __boost_mem) const
```
`__boost_mem` is unused variable

danhoeflinger replied:
This is to provide a unified interface with the other specialization so that the bin hash can be treated generically in the kernel.  Evenly divided bins do not take advantage of extra SLM memory, but custom range bins do, to store the `__boundaries`.

mmichel11 replied:
Would it be worth it to decorate `__boost_mem` as `[[maybe_unused]]`, so the user does not receive any additional warnings when compiling with `-Wunused-parameter`?

danhoeflinger replied:
Considering it further, it makes more sense to make it an unnamed parameter for this overload.  I think this makes it clear that it is intentionally unused and avoids those warnings as well.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+
+    template <typename _T2>
+    inline bool
+    is_valid(_T2&& __value, void* __boost_mem) const
```
`__boost_mem` is unused variable

danhoeflinger replied:
Same as above.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+    auto __f = oneapi::dpl::__internal::fill_functor<__global_histogram_type>{__global_histogram_type{0}};
+    //fill histogram bins with zeros
+    auto init_e = oneapi::dpl::__par_backend_hetero::__parallel_for(
+        ::std::forward<_ExecutionPolicy>(exec), unseq_backend::walk_n<_ExecutionPolicy, decltype(__f)>{__f}, __num_bins,
```
We should not forward `exec` here since it will likely be used again later in this function.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+
+    auto __f = oneapi::dpl::__internal::fill_functor<__global_histogram_type>{__global_histogram_type{0}};
+    //fill histogram bins with zeros
+    auto init_e = oneapi::dpl::__par_backend_hetero::__parallel_for(
```
Since we are initializing to all zeros, could it potentially be faster to initialize all bytes as 0 with `queue::memset` instead of a parallel_for with a fill functor?

danhoeflinger replied:
This was my original approach in prototyping, but since the API has user supplied **iterator** input,  I think this may be the best we can do.  
I user could provide a `permutation_iterator` in for instance as the output histogram.

adamfidel commented on include/oneapi/dpl/internal/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,129 @@
+    inline ::std::uint32_t
+    get_bin(_T2&& __value) const
+    {
+        return (::std::upper_bound(__boundaries.begin(), __boundaries.end(), ::std::forward<_T2>(__value)) -
```
I'm not sure what the expected type for `_Range` is here, but `::std::distance` might be better here in the cases where the iterator type of `_Range` is not random access.

This might also apply to the use of `operator[]` in the `is_valid` method below.

adamfidel commented on include/oneapi/dpl/internal/histogram_impl.h:
```diff
@@ -0,0 +1,117 @@
+                    _RandomAccessIterator2 __histogram_first, const _Size& __num_bins, _IdxHashFunc __func,
+                    _Range&&... __opt_range)
+{
+    assert(false && "Histogram API is not currently supported for host parallel policies");
```
This might work better as a `static_assert`, since any calls to this overload with a host execution policy is known at compile time to be an error.

danhoeflinger replied:
Aha, thanks for pointing that out.  I'll try it.  I agree a static assert is far superior if we can get it to work.

danhoeflinger replied:
Yes, this seems to have fixed it (although the CI seems to be having other problems unrelated to this PR now...)

adamfidel commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+    {
+        ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+        ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+        ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__bin_hash.__boundaries.size(), __gSize);
```
Is there any guarantee that this value will fit within 8 bits? I can imagine a scenario with a small work group size (64) and a decent amount of bins (32,768) which would cause this value to not be represented in a `uint8_t`.

MikeDvorskiy commented on include/oneapi/dpl/internal/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    _Range
+    get_range()
```
`get_range() const` ?

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+template <::std::uint16_t __iters_per_work_item, typename _ExecutionPolicy, typename _Iter1, typename _Iter2,
+          typename _Size, typename _IdxHashFunc, typename... _Range>
+inline auto
+__parallel_histogram_sycl_impl(_ExecutionPolicy&& __exec, _Iter1 __first, _Iter1 __last, _Iter2 __histogram_first,
```
1. we already mentioned "sycl" in the filename, so probably we dont need duplicate it in the name of the function(?)
2. Also I'm confused with argument types - there are "iterators" and "ranges" at the same time... Before we have or an iterator based signature or a range based signature for an algorithm.

danhoeflinger replied:
1.  OK, changed the name.

2.  The public API is iterators only.  Previously, this code location was in the middle of conversion of different inputs to sycl ranges, which occurred at different times because of the unification of the 2 APIs.  
    
	To make this more consistent, I moved the sycl range transformation for the always required input elements and and output bins up the stack to `__parallel_histogram`, which makes each level either all ranges or all iterators, and still allows it to be unified for both APIs. Thanks, this refactor improved the code.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+    if (__func.is_valid(__x))
+    {
+        _BinIdxType c = __func.get_bin(__x, __boost_mem);
+        __histogram[c]++;
```
In given case, I trust to the modern compiler that it makes in-place increment here.... But, several years was a rule to have pre-increment to avoid a temporary variable creation (even in such case  - a kind of style..... I don't insist).

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+inline auto
+__histogram_general_private_global_atomics(_ExecutionPolicy&& __exec, const sycl::event& __init_e,
+                                           ::std::uint16_t __work_group_size, _Range1&& __input, _Range2&& __bins,
+                                           const _Size& __num_bins, _IdxHashFunc __func, _Range3&&... __opt_range)
```
It looks that we have "redundant" signature here, in meaning `_Range2&& __bins` already has information about number  of bins or Did I miss something?

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,547 @@
+
+template <typename _BinIdxType, typename _Iter1, typename _HistReg, typename _BinFunc>
+inline void
+__accum_local_register_iter(const _Iter1& __in_acc, const ::std::size_t& __index, _HistReg* __histogram,
```
`const _Iter1& __in_acc`
Is it iterator or accessor?

danhoeflinger replied:
Believe it or not... it should be a Range.  Fixed.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_impl.h:
```diff
@@ -0,0 +1,117 @@
+          const _Size& __num_bins, const _T& __first_bin_min_val, const _T& __last_bin_max_val,
+          _RandomAccessIterator2 __histogram_first)
+{
+    __internal::__histogram_impl(::std::forward<_ExecutionPolicy>(exec), __first, __last, __histogram_first, __num_bins,
```
The reason is not satisfying.  It is because the public API changed later in the process to place the output at the end, and the details did not.  

I've fixed it so the details match the public interface up until converting to ranges, thanks.

danhoeflinger commented:
@MikeDvorskiy @SergeyKopienko @mmichel11 @adamfidel and all:

@rarutyun and I discussed a possible change of the API to something like this:

```
template <typename _Size, typename _ValueType>
oneapi::dpl::even_histogram_config
make_histogram_config(_Size num_bins, _ValueType first_bin_min_val, _ValueType last_bin_max_val);

template <typename _RandomAccessIterator>
oneapi::dpl::custom_range_histogram_config
make_histogram_config(_RandomAccessIterator boundary_first, _RandomAccessIterator boundary_last);


template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _HistogramConfig, typename _RandomAccessIterator2>
oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _RandomAccessIterator2>
histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 first, _RandomAccessIterator1 last, _HistogramConfig histogram_config,
          _RandomAccessIterator2 histogram_first);
```
where `_HistogramConfig` must be `even_histogram_config` or `custom_range_histogram_config`.

Calls could look like:
``` 
histogram(exec, input_first, input_last, make_histogram_config(10, 0, 100), output_first);

auto hist_config = make_histogram_config(boundary_first, boundary_last);

histogram(exec, input_first, input_last, hist_config, output_first);
```

We want to raise it to the group to see if there is any strong support for such a change of API.  A brief listing of pros and cons we see:

Pros:
* Allows fewer overloads for supporting a mixture of ranges and iterators (for potential future range APIs)
* Allows easier re-use of a histogram "configuration" with different inputs / outputs

Cons:
* Departure from other std parallel APIs, with addition of separated configuration 
* Makes migration more complicated within SYCLomatic
* Lose ability to connect the type of bin range min and max to the value type of the input sequence for evenly divided API

We have not found overwhelming motivation for this change, but I think it is worth raising as an option.  
If there is not strong support for this change, we will leave the API as it is in the PR currently.

Please share your thoughts.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_extension_defs.h:
```diff
@@ -0,0 +1,47 @@
+
+// Only supported for dpcpp backend
+#ifdef _ONEDPL_BACKEND_SYCL
+
```
@MikeDvorskiy @SergeyKopienko @mmichel11 @adamfidel and all:

@rarutyun and I discussed a possible change of the API to something like this:

```
template <typename _Size, typename _ValueType>
oneapi::dpl::even_histogram_config
make_histogram_config(_Size num_bins, _ValueType first_bin_min_val, _ValueType last_bin_max_val);

template <typename _RandomAccessIterator>
oneapi::dpl::custom_range_histogram_config
make_histogram_config(_RandomAccessIterator boundary_first, _RandomAccessIterator boundary_last);


template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _HistogramConfig, typename _RandomAccessIterator2>
oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _RandomAccessIterator2>
histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 first, _RandomAccessIterator1 last, _HistogramConfig histogram_config,
          _RandomAccessIterator2 histogram_first);
```
where `_HistogramConfig` must be `even_histogram_config` or `custom_range_histogram_config`.

Calls could look like:
``` 
histogram(exec, input_first, input_last, make_histogram_config(10, 0, 100), output_first);

auto hist_config = make_histogram_config(boundary_first, boundary_last);

histogram(exec, input_first, input_last, hist_config, output_first);
```

We want to raise it to the group to see if there is any strong support for such a change of API.  A brief listing of pros and cons we see:

Pros:
* Allows fewer overloads for supporting a mixture of ranges and iterators (for potential future range APIs)
* Allows easier re-use of a histogram "configuration" with different inputs / outputs

Cons:
* Departure from other std parallel APIs, with addition of separated configuration 
* Makes migration more complicated within SYCLomatic
* Lose ability to connect the type of bin range min and max to the value type of the input sequence for evenly divided API

We have not found overwhelming motivation for this change, but I think it is worth raising as an option.  
If there is not strong support for this change, we will leave the API as it is in the PR currently.

Please share your thoughts.

rarutyun commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,671 @@
+    auto __boundary_buf = __keep_boundaries(__range_to_upg.begin(), __range_to_upg.end());
+    auto __boundary_view = __boundary_buf.all_view();
+    auto __bin_hash = oneapi::dpl::__internal::__custom_range_binhash(__boundary_view);
+    return __parallel_histogram_impl<__iters_per_work_item>(
```
You have "interesting" `Range3` optional parameter for the overload above, but I cannot see how it's further used except `__require_access` function

danhoeflinger replied:
Yes, the `Range3` sequence is used from within the "custom range" binhash structure to hold the boundary sequence (but not used for the evenly divided binhash structure).  This is why it is optional and passed along to the kernels to be required.  It is used by the binhash structure internally (if applicable).

I should be able to work out a way to minimize the number of levels of indirection this optional parameter is passed through by encoding / using some information in the binhash function closer to the kernel launch.  I'll explore that, as it would improve the readability of this code.

danhoeflinger replied:
I've fixed this to fold completely into the binhash structure to be required in the submit call.  In hindsight, this seems the obvious way to do it, and simplifies the code a lot, thanks.

rarutyun reviewed (commented):
First portion of comments.

Didn't look at tests at all. Still in progress with thinking about the implementation

danhoeflinger commented on include/oneapi/dpl/internal/histogram_impl.h:
```diff
@@ -0,0 +1,116 @@
+                    _Size __num_bins, _IdxHashFunc __func, _RandomAccessIterator2 __histogram_first)
+{
+    //If there are no histogram bins there is nothing to do
+    if (__num_bins > 0)
```
I think this is what we want.  The semantics are that if elements don't fall into bins (if they are out of range), they are silently skipped.  I see this case as all input elements are skipped because they don't map to any existing bin.

I'd be open to some error message instead, but I think this makes sense.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_impl.h:
```diff
@@ -0,0 +1,116 @@
+#if _ONEDPL_BACKEND_SYCL
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _IdxHashFunc,
+          typename _RandomAccessIterator2>
+inline oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<_ExecutionPolicy>::type>
```
I don't have a good justification, I'll remove them.  Thanks.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_impl.h:
```diff
@@ -0,0 +1,116 @@
+__pattern_histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _Size __num_bins, _IdxHashFunc __func, _RandomAccessIterator2 __histogram_first)
+{
+    static_assert(sizeof(_Size) == 0 /*false*/, "Histogram API is not currently supported for host parallel policies");
```
The intention is to add this implementation at a later time.
For now, I think that this makes things most clear that we are intentionally leaving out the CPU implementation of this API, rather than just a compilation error about a missing function definition for `__pattern_histogram` if a user tries to call it with with a host parallel policy.
  
I don't feel very strongly on this topic though, so I'm open to other options. Do you have a preference to just remove this definition or explicitly delete it?

dcbenito reviewed (changes requested):
Since these are examples and not headers, I would change to sentence case:
Evenly divided bins:
...
Custom range bins:
...

danhoeflinger commented:
After an offline discussion with @rarutyun and some thought on my own, I've made some refactors to address the feedback and generally improve the code.  I will list the larger changes here (minor stuff can be found in the change log):

1. Adjusted SLM wrapper to have less responsibility and to be created within the kernel (no longer needs to be device copyable). This allows it to be initialized with the SLM local accessor, rather than passing in the local accessor to each call.
2. Shifted some of that responsibility from SLM wrapper to the `__binhash_manager_...` to `require_access`, provide the type and size of the required SLM allocation.
3. Adjusted partial template specialization to instead be simple function overloads for what is now `__make_binhash_manager`.
4. Removed unnecessary `_impl` layer near the public interface.
5. Consolidated duplicated implementations of custom range binhash functionality to static helper functions which can be used both by the top level binhash and the SLM wrapper, with different boundary data.

The kernels remain unchanged and these changes don't have an affect on performance outside the margin of error.

I've also added the documentation of the new API.

adamfidel commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,681 @@
+struct __binhash_manager_with_buffer : __binhash_manager_base<_BinHash>
+{
+    using BaseType = __binhash_manager_base<_BinHash>;
+    using extra_memory_type = typename _BinHash::range_value_type;
```
Some typedefs in this file are in `snake_case` and others are in `PascalCase`. In this particular instance, we are using different case conventions for two lines that are right next to eachother.

Should we decide on a single style for these typedefs?

In either case, both of these typedefs probably should be prefixed with an underscore.

danhoeflinger replied:
Thanks, this is a good point.  I've improved the naming of some of these, and I am using snake case with the leading `_`.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,681 @@
+// -*- C++ -*-
+//===-- parallel_backend_sycl_reduce.h --------------------------------===//
```
This file name comment should be `parallel_backend_sycl_histogram.h`

danhoeflinger replied:
good catch, fixed.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,681 @@
+}
+
+template <typename _BinHash>
+struct __binhash_manager_base
```
I think expressing of the polymorphic behavior is so "complicated" and tricky here. 
We have a "base" class (which does nothing)  with a couple of dummy methods and a derived class which inherits just  `_BinHash ` field.
But by fact we have two class specializations. 
I believe the code may be reduced and made more shorter and cleared.

We have already two specializations  - `__evenly_divided_binhash` and `__custom_range_binhash`...
We may add a general bool property, for example, to "inform" kernel whether  we need SLM or don't.

danhoeflinger replied:
I experimented with this using template specialization, and was unhappy with the result, and went back to the polymorphic approach.  You can see the difference between that attempt and the current PR at
https://github.com/oneapi-src/oneDPL/pull/1273/commits/efbb843feebe50f2f44ab79752ef9d80939174a0 .  I'm sure there is a better way to do it with the template specialization way, without having to explicitly specify the types, but I'm not finding it.

I think I prefer the polymorphic approach because we *are* building functionality on, by making changes to accommodate extra memory that needs a call to `require_access()`, has a size for SLM allocations, and requires a `sycl::buffer` lifetime extender.  Both need to return a valid bin hash function to be used in the kernel.

As far as the bool property for SLM vs no SLM (I assume instead of the zero SLM memory return), we should discuss this further offline.  I would prefer the kernels not need to know the details of the binhash function and be written generically.  I think this makes the kernels cleaner, and doesn't need branching to special case for individual binhash functions.  I'm open to a discussion but there is some nuance that needs to be worked through with code examples.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,681 @@
+{
+    auto __range_to_upg = __bin_hash.get_range();
+    auto __keep_boundaries =
+        oneapi::dpl::__ranges::__get_sycl_range<oneapi::dpl::__par_backend_hetero::access_mode::read,
```
I think there is some confusion here.  There are 2 or 3 sequences which require `__get_sycl_range` depending on the API which was called.
1) input elements to be processed
2) output bins to use for the histogram counts
3) boundary data to determine bins **only for the custom bin API, not evenly divided API**

For each of them, they are only applying `__get_sycl_range` once.  In this comment you are conflating the output bins with the boundaries used to define the bins.  This function processes the boundaries, (and only is called when `__custom_range_binhash` is the hash function.  The linked code processes the output histogram code, and is always called.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,681 @@
+        using _extra_memory_type = typename _BinHashMgr::_extra_memory_type;
+        auto _device_copyable_func = __binhash_manager.get_device_copyable_binhash();
+
+        ::std::size_t __extra_SLM_elements = __binhash_manager.get_required_SLM_elements();
```
`__binhash_manager_base` is used together with binhash functions which have no extra sequence of memory which requires a buffer to represent them (`evenly_divided_binhash`).  You link to the `custom_range_binhash` specialization of `__binhash_SLM_wrapper`, which has an extra sequence on memory (the boundaries) which need to be held in a buffer, and will therefore us `__binhash_manager_with_buffer` instead.
No out of range access will occur here.

It would be nice if `binhash_manager_...` and `__binhash_SLM_wrapper` could be combined, but the first needs to be set up in host code and keep alive a `sycl::buffer` for the duration of the kernel, and the other needs to be created in the kernel to be able to be created with an SLM accessor, and accelerate the kernel without passing around SLM accessors to each call.

danhoeflinger replied:
remove me

danhoeflinger commented on include/oneapi/dpl/internal/histogram_impl.h:
```diff
@@ -0,0 +1,95 @@
+    ::std::ptrdiff_t num_bins = boundary_last - boundary_first - 1;
+    auto boundary_view = oneapi::dpl::__ranges::guard_view<_RandomAccessIterator2>(boundary_first, boundary_last);
+    __internal::__pattern_histogram(::std::forward<_ExecutionPolicy>(exec), first, last, num_bins,
+                                    __internal::__custom_range_binhash{boundary_view}, histogram_first);
```
We do `__get_sycl_range` for the boundaries, in the `__make_binhash_manager` function.  Perhaps we just need to extract the `__get_sycl_range` calls up a level or two, outside of the sycl backend to be able to support a potential asynchronous version of this pattern in the same way as the rest of the library.
I will investigate this tomorrow, and see if this can be done reasonably.  For now, without such an asynchronous version, I don't see that there is a correctness or stability issue.

I use ranges within the binhash functions so that we can use one implementation for CPU code and the SYCL kernels after `__get_sycl_range`.  I just defer turning this range into a sycl range until it is in the sycl backend in the current code.

MikeDvorskiy replied:
> I don't see that there is a correctness or stability issue.

It is a code design issue (for SYCL backend).

danhoeflinger replied:
As mentioned above, I've added this middle layer and consolidated the `__get_sycl_range` calls there, above the `__parallel_histogram` level.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_impl.h:
```diff
@@ -0,0 +1,95 @@
+#include "../pstl/iterator_impl.h"
+
+#if _ONEDPL_HETERO_BACKEND
+#    include "oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h"
```
I believe that this should all be SYCL, rather than hetero.  I will resolve this tomorrow.

danhoeflinger replied:
Thanks, I've fixed this by adding the separated middle hetero layer `include/oneapi/dpl/pstl/hetero/histogram_impl_hetero.h`

danhoeflinger commented on include/oneapi/dpl/internal/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,131 @@
+#ifndef _ONEDPL_HISTOGRAM_BINHASH_UTILS_H
+#define _ONEDPL_HISTOGRAM_BINHASH_UTILS_H
+
+#include "../pstl/utils_ranges.h"
```
I've moved the range implementation into the SYCL backend, and have an iterator version now at this location.  They both use the same standalone helper function to implement the `get_bin()` function.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,681 @@
+
+template <typename _BinIdxType, typename _Range, typename _HistReg, typename _BinFunc>
+void
+__accum_local_register_iter(_Range&& __input, const ::std::size_t& __index, _HistReg* __histogram, _BinFunc __func)
```
Thanks, I've taken your suggestions for 1,2,3.  I would prefer to leave it as a separate helper function so it can be reused rather than repeated.  It also allows us to mirror the structure of the other kernels.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,131 @@
+
+    template <typename _T2>
+    ::std::uint32_t
+    get_bin(_T2&& __value) const
```
Thanks, I've taken both of these suggestions.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,131 @@
+struct __custom_range_binhash
+{
+    using _range_value_type = oneapi::dpl::__internal::__value_t<_Range>;
+    _Range __boundaries;
```
I've updated this to use iterators at this location.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,131 @@
+        return is_valid_helper(__boundaries[0], __boundaries[__boundaries.size() - 1], ::std::forward<_T2>(__value));
+    }
+
+    _Range
```
I've removed this function entirely.  The only place we needed it, we are specializing specifically to this structure, so I just use the iterators directly now.

danhoeflinger commented on include/oneapi/dpl/internal/histogram_impl.h:
```diff
@@ -0,0 +1,95 @@
+          _RandomAccessIterator3 histogram_first)
+{
+    ::std::ptrdiff_t num_bins = boundary_last - boundary_first - 1;
+    auto boundary_view = oneapi::dpl::__ranges::guard_view<_RandomAccessIterator2>(boundary_first, boundary_last);
```
Thanks, I've incorporated these changes.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,681 @@
+
+template <typename _ExecutionPolicy, typename _Iter1, typename _Size, typename _IdxHashFunc, typename _Iter2>
+void
+__parallel_histogram(_ExecutionPolicy&& __exec, _Iter1 __first, _Iter1 __last, _Size __num_bins, _IdxHashFunc __func,
```
I think it would be best to add histogram to the experimental async and range APIs in a follow up PR.  

Unless you see this as a requirement for approval for this PR, I will make an issue to add this as a follow up PR.

danhoeflinger commented on test/support/histogram_serial_impl.h:
```diff
@@ -0,0 +1,79 @@
+histogram_sequential(_InputIterator1 __first, _InputIterator1 __last, _Size __num_bins,
+                     _T __first_bin_min_val, _T __last_bin_max_val, _OutputIterator __histogram_first)
+{
+    for (int bin = 0; bin < __num_bins; bin++)
```
Fixed.

danhoeflinger commented on test/support/histogram_serial_impl.h:
```diff
@@ -0,0 +1,79 @@
+                     _InputIterator2 __boundary_last, _OutputIterator __histogram_first)
+{
+    int num_bins = (__boundary_last - __boundary_first) - 1;
+    for (int bin = 0; bin < num_bins; bin++)
```
Fixed.

MikeDvorskiy commented on include/oneapi/dpl/pstl/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,120 @@
+    ::std::int32_t
+    get_bin(const _T2& __value) const
+    {
+        int ret = -1;
```
(I know that I "pushed" you to introduce "<0" for invalid values.)
In the structure we have 3 different types for one essence - `::std::uint32_t`,  `::std::int32_t` and `int`. I think it should be arranged into one type `_BinIndex` for example, which may be passed as template parameter.

Moreover, it makes sense to add  asserts (somewhere in the impl) for "exotic" cases, when we may get integer overflow with number  of bins, 
static assert which checks `std::is_signed<_BinIndex>`
and runtime assert which checks `__num_bins <= std::numeric_limits::max<_BinIndex>`

danhoeflinger replied:
I agree with the comment in that these should all be unified.  However, I don't see the value in allowing a template parameter dictate the type.  I think it should always just be `::std::int32_t`.  I've added the runtime asserts on creation of the binhash functions.  The static assert only makes sense with the template parameter.

MikeDvorskiy commented on include/oneapi/dpl/pstl/histogram_impl.h:
```diff
@@ -0,0 +1,77 @@
+histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 first, _RandomAccessIterator1 last, _Size num_bins,
+          _ValueType first_bin_min_val, _ValueType last_bin_max_val, _RandomAccessIterator2 histogram_first)
+{
+    __internal::__pattern_histogram(
```
Is it better (and we try to do that everywhere in oneDPL) to use fully qualified names => `oneapi::dpl::__internal::__pattern_histogram`

danhoeflinger replied:
added

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/histogram_impl_hetero.h:
```diff
@@ -0,0 +1,174 @@
+            auto __input_buf = __keep_input(__first, __last);
+
+            __parallel_histogram(::std::forward<_ExecutionPolicy>(__exec), __init_event, __input_buf.all_view(),
+                                 ::std::move(__bins), ::std::move(__binhash_manager))
```
Are you sure that you want moving `__binhash_manager`?
The fact is it contains `__buffer_lifetime_holder`. M.b. we should keep the lifetime in this scope? 
I am asking to be sure that we don't lose "control" here.

MikeDvorskiy commented on include/oneapi/dpl/pstl/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    template <typename _T2>
+    ::std::int32_t
```
the same comment about `::std::int32_t/int`

MikeDvorskiy commented on include/oneapi/dpl/pstl/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,120 @@
+};
+
+template <typename _BoundaryIter, typename _T2, typename _T3>
+int
```
a return type is `int`, but ::std::distance returns a difference type...

danhoeflinger replied:
now with the assert, we ensure we wont overflow.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,564 @@
+    __custom_boundary_range_binhash(_Range __boundaries_) : __boundaries(__boundaries_) {}
+
+    template <typename _T2>
+    ::std::int32_t
```
=> auto ?
(or some _BinIndex)

danhoeflinger replied:
I've updated all these `get_bin` returns to be `auto` where it makes sense.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,564 @@
+
+    template <typename _T2>
+    ::std::int32_t
+    get_bin(_T2&& __value) const
```
Do we really need `_T2&&` (and `::std::forward<_T2>`)?

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,564 @@
+
+    template <typename _T>
+    ::std::int32_t
+    get_bin(_T&& __value) const
```
Do we really need _T2&& (and ::std::forward<_T2>)?

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,564 @@
+    }
+
+    template <typename _T>
+    ::std::int32_t
```
-> auto ?
(or some _BinIndex)

danhoeflinger replied:
fixed.

MikeDvorskiy reviewed (commented):
Code design with split into parts "host/hetero/SYCL specific" is getting better.
Also I've added some new comments.

danhoeflinger commented on include/oneapi/dpl/pstl/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,120 @@
+    int ret = -1;
+    if (__value >= __min && __value < __max)
+    {
+        ret = ::std::distance(__first, ::std::upper_bound(__first, __last, ::std::forward<_T2>(__value))) - 1;
```
fixed.

danhoeflinger commented on include/oneapi/dpl/pstl/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,120 @@
+
+    template <typename _T2>
+    ::std::int32_t
+    get_bin(_T2&& __value) const
```
changed to pass by value.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h:
```diff
@@ -0,0 +1,564 @@
+
+    template <typename _T>
+    ::std::int32_t
+    get_bin(_T&& __value) const
```
changed to pass by value

danhoeflinger commented on include/oneapi/dpl/pstl/histogram_impl.h:
```diff
@@ -0,0 +1,77 @@
+          _RandomAccessIterator3 histogram_first)
+{
+    ::std::ptrdiff_t num_bins = boundary_last - boundary_first - 1;
+    __internal::__pattern_histogram(::std::forward<_ExecutionPolicy>(exec), first, last, num_bins,
```
added

danhoeflinger reviewed (commented):
Addressed the new comments other than the return type comments for get_bin.  Working on that now.

danhoeflinger commented on include/oneapi/dpl/pstl/histogram_extension_defs.h:
```diff
@@ -0,0 +1,47 @@
+{
+
+// Only supported for dpcpp backend
+#ifdef _ONEDPL_BACKEND_SYCL
```
yes good catch.  Fixed.

MikeDvorskiy commented on include/oneapi/dpl/pstl/histogram_binhash_utils.h:
```diff
@@ -0,0 +1,123 @@
+        ::std::int32_t ret = -1;
+        if ((__value >= __minimum) && (__value < (__minimum + __range_size)))
+        {
+            ret = ((::std::uint64_t(__value) - __minimum) * ::std::uint64_t(__num_bins)) / __range_size;
```
The first operand `((::std::uint64_t(__value) - __minimum)` is correct?
Probably should be  `::std::uint64_t(__value - __minimum)` ?

danhoeflinger replied:
Yes, fixed.  I've also improved the testing a bit to try to have covered that case.  However, I think that the code was underflowing uint64 in a way to end up "accidentally" at the correct result anyway after the subtraction.

MikeDvorskiy commented on include/oneapi/dpl/pstl/hetero/histogram_impl_hetero.h:
```diff
@@ -0,0 +1,174 @@
+                    _Size __num_bins, _BinHash&& __func, _RandomAccessIterator2 __histogram_first)
+{
+    //If there are no histogram bins there is nothing to do
+    if (__num_bins > 0)
```
Just minor comment for the future.
There is recommendation to keep the sorter (or no op) code branch as first branch of "if else" operator. It improves readability (easier to control code branching, no indentation for many lines), and sometimes makes smaller the file size.

MikeDvorskiy reviewed (dismissed):
At the moment all issue which I noticed were fixed.
Some code improvements with code design perspective may be done in the future.
So, I think we finalized the PR as far as we can for today.

MikeDvorskiy reviewed (approved):
Re- approve after adding `oneapi::dpl::__internal::__pstl_upper_bound` (index version of std::upper_bound).
